sway/commands/mirror.c

```diff
@@ -0,0 +1,297 @@
+	params.flavour = SWAY_MIRROR_FLAVOUR_ENTIRE;
+
+	if (!build_dst_params(&params, argv[0], argv[1])) {
+		return cmd_results_new(CMD_FAILURE, params_failure_message);
+	}
+
+	if (!build_src_params(&params, argv[2])) {
+		return cmd_results_new(CMD_FAILURE, params_failure_message);
+	}
+
+	if (params.output_src == params.wlr_params.output_dst) {
+		return cmd_results_new(CMD_FAILURE, "src and dst must be different");
+	}
+
+	if (argc == 4) {
+		if (strcmp(argv[3], "show_cursor") != 0) {
+			return cmd_results_new(CMD_FAILURE, usage);
+		}
+		params.wlr_params.overlay_cursor = true;
+	}
+
+	if (!mirror_create(&params)) {
+		return cmd_results_new(CMD_FAILURE, "Mirror failed to start, check logs.");
+	}
+
+	return cmd_results_new(CMD_SUCCESS, NULL);
+}
+
+struct cmd_results *cmd_mirror_start_box(int argc, char **argv) {
+	const char usage[] = "Expected 'mirror start box "
+		"<dst_name> <full|aspect|center> "
+		"<x> <y> <width> <height> [show_cursor]'";
+
+	if (argc < 6 || argc > 7) {
+		return cmd_results_new(CMD_INVALID, usage);
+	}
+
+	struct sway_mirror_params params = { 0 };
+	params.flavour = SWAY_MIRROR_FLAVOUR_BOX;
+
+	if (!build_dst_params(&params, argv[0], argv[1])) {
+		return cmd_results_new(CMD_FAILURE, params_failure_message);
+	}
+
+	char *end;
+	params.box.x = strtol(argv[2], &end, 10);
+	if (end[0] != '\0' || params.box.x < 0) {
+		return cmd_results_new(CMD_FAILURE, "Invalid x '%s'.", argv[2]);
+	}
+	params.box.y = strtol(argv[3], &end, 10);
+	if (end[0] != '\0' || params.box.y < 0) {
+		return cmd_results_new(CMD_FAILURE, "Invalid y '%s'.", argv[3]);
+	}
+	params.box.width = strtol(argv[4], &end, 10);
+	if (end[0] != '\0' || params.box.width < 1) {
+		return cmd_results_new(CMD_FAILURE, "Invalid width '%s'.", argv[4]);
+	}
+	params.box.height = strtol(argv[5], &end, 10);
+	if (end[0] != '\0' || params.box.height < 1) {
+		return cmd_results_new(CMD_FAILURE, "Invalid height '%s'.", argv[5]);
+	}
+
+	// find the output for the origin of the box
+	params.output_src = wlr_output_layout_output_at(root->output_layout,
+			params.box.x, params.box.y);
+	if (!params.output_src) {
+		return cmd_results_new(CMD_FAILURE, "Box not on any output.");
+	}
+
+	// box cannot be on the dst
+	if (params.output_src == params.wlr_params.output_dst) {
+		return cmd_results_new(CMD_FAILURE, "Box on dst.");
+	}
+
+	// convert to local output coordinates, ensuring within output_src
+	if (!mirror_layout_box_within_output(&params.box, params.output_src)) {
+		return cmd_results_new(CMD_FAILURE, "Box covers multiple outputs.");
+	}
+
+	if (argc == 7) {
+		if (strcmp(argv[6], "show_cursor") != 0) {
+			return cmd_results_new(CMD_FAILURE, usage);
+		}
+		params.wlr_params.overlay_cursor = true;
+	}
+
+	if (!mirror_create(&params)) {
+		return cmd_results_new(CMD_FAILURE, "Mirror failed to start, check logs.");
+	}
+
+	return cmd_results_new(CMD_SUCCESS, NULL);
+}
+
+struct cmd_results *cmd_mirror_start_container(int argc, char **argv) {
```

Comment:
subcommands could be used, however may be impractical as these starts share a lot.